imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+        compression_width = int("".join([str(i) for i in bit_array[:6]]), 2)
+        has_range_data_section = int(str(bit_array[6]), 2)
+
+        # The full vector includes 3 values of compression_width bits, and excludes
+        # range.
+        uncompressed_vector_size = compression_width * 3
+        # plus 8 to get past the compression width and range data section
+        first_vector_width = uncompressed_vector_size + 8 + 2
+
+        first_vector = MagL1a.unpack_one_vector(
+            bit_array[8:first_vector_width], compression_width, True
+        )
+
+        end_vector = (
+            len(bit_array)
+            - (primary_count + secondary_count) * 2 * has_range_data_section
+        )
+
+        # Cut off the first vector width and the end range data section if it exists.
+        vector_bits = bit_array[first_vector_width - 1 : end_vector]
+
+        # Shift the bit array over one to the left, then sum them up. This is used to
+        # find all the places where two 1s occur next to each other, because the sum
+        # will be 2 for those indices.
+        # For example: [0 0 1 0 1 1] + [1 0 0 1 0 1] = [1 0 1 1 1 2], so the last index
+        # has 2 ones in a row.
+        # The first bit is invalid, so we remove it at the end.
+        sequential_ones = np.where(
+            np.add(
+                vector_bits,
+                np.roll(vector_bits, 1),
+            )[1:]
+            == 2
+        )[0]
+
+        # The first bit is only needed for the np.roll step, so now we remove it.
+        # we are left with compressed primary vectors, and all the secondary vectors.
+        vector_bits = vector_bits[1:]
+        # print(f"Vector bits: {vector_bits}")
+        # which indices within vector_bits are the end of a vector
+        primary_boundaries = [sequential_ones[0] + 1]
+        secondary_boundaries = []
+        vector_count = 1
+        end_primary_vector = 0
+        for seq_val in sequential_ones:
+            if vector_count > primary_count + secondary_count:
+                break
+            # Add the end indices of each primary vector to primary_boundaries
+            # If we have 3 ones in a row, we should skip that index
+            if vector_count < primary_count and (
+                seq_val - primary_boundaries[-1] > 0
+            ):
+                primary_boundaries.append(seq_val + 1)
+
+                # 3 boundaries equal one vector
+                if len(primary_boundaries) % 3 == 0:
+                    vector_count += 1
+                    # If the vector length is >60 bits, we switch to uncompressed.
+                    # So we skip past all the remaining seq_ones.
+                    if (
+                        (len(primary_boundaries) > 4)
+                        and (primary_boundaries[-1] - primary_boundaries[-4] > 60)
+                        or (vector_count == 2 and primary_boundaries[-1] > 60)
```

Comment:
This line caused me much trouble. The beginning and ends of the sequence are kind of weird, since each index indicates the end of the value. This means we need to compare the most recent value to 3 indices ago - except for the first vector, which does not have  3 prior indices, only 2. So in that case we just look at the index of the last value to see if it's > 60. 

I will update the comment to explain this a bit more, thanks for calling it out.